File: latexdraw-research/GUImutants/mutant14/net.sf.latexdraw.mutant14/src/main/net/sf/latexdraw/actions/shape/DistributeShapes.scala
package net.sf.latexdraw.actions.shape

import org.malai.action.Action
import org.malai.undo.Undoable
import net.sf.latexdraw.actions.Modifying
import net.sf.latexdraw.actions.ShapeAction
import net.sf.latexdraw.glib.models.interfaces.IShape
import java.awt.geom.Rectangle2D
import org.malai.mapping.MappingRegistry
import net.sf.latexdraw.glib.views.Java2D.interfaces.IViewShape
import net.sf.latexdraw.glib.models.interfaces.IGroup
import scala.collection.JavaConversions._
import scala.collection.mutable.MutableList
import scala.collection.mutable.ListBuffer
import net.sf.latexdraw.util.LNumber
import net.sf.latexdraw.glib.models.interfaces.IPoint

/**
 * This enumeration describes the different possible alignment types.
 */
object DistributionType extends Enumeration {
	type DistributionType = Value
	val vertBot, vertTop, vertMid, vertEq, horizLeft, horizRight, horizMid, horizEq = Value

	def isVertical(distrib:DistributionType) = distrib==vertBot || distrib==vertTop || distrib==vertMid || distrib==vertEq
}

import DistributionType._

/**
 * This action distributes the provided shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2013-31-03<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class DistributeShapes extends Action with ShapeAction[IGroup] with Undoable with Modifying {
	/** The reference border that must bounds the shapes to align. */
	var _border : Rectangle2D = null

	/** The views corresponding to the shapes to align. */
	val _views = new ListBuffer[IViewShape]()

	/** The alignment to perform. */
	var _distribution : DistributionType = null

	/** The former positions of the shapes to align. Used for undoing. */
	val _oldPositions = new ListBuffer[IPoint]()


	override protected def doActionBody() {
		var v : IViewShape = null
		shape.get.getShapes.foreach{sh=>
			// Because the views have already computed the border of their shape, we get the corresponding views.
			_views += MappingRegistry.REGISTRY.getTargetFromSource(sh, classOf[IViewShape])
			// Saving the old position of the shape for undoing.
			_oldPositions += sh.getTopLeftPoint
		}
		redo
	}


	override def canDo = _shape.isDefined && !_shape.get.isEmpty && _border!=null && _distribution!=null


	override def undo() {
		var pt : IPoint = null
		var pos : Int = 0
		var oldPt : IPoint = null

		shape.get.getShapes.foreach{sh=>
			// Reusing the old position.
			pt = sh.getTopLeftPoint
			oldPt = _oldPositions(pos)
			if(!pt.equals(oldPt))
				sh.translate(oldPt.getX-pt.getX, oldPt.getY-pt.getY)
			pos += 1
		}
		shape.get.setModified(true)
	}


	/**
	 * Distributes at equal distance between the shapes.
	 */
	protected def distributeEq() {
		var sortedSh= new ListBuffer[IViewShape]()
		var mins	= new ListBuffer[Double]()
		var maxs	= new ListBuffer[Double]()
		var i : Int = 0
		val shapes = shape.get.getShapes

		_views.foreach{view =>
			val coord = _distribution match {
				case DistributionType.horizEq => view.getBorder.getMinX
				case DistributionType.vertEq => view.getBorder.getMinY
			}

			i = mins.indexWhere{value=> coord<value}

			if(i == -1) {
				sortedSh+=view
				mins+=coord
				if(_distribution==DistributionType.horizEq)
					 maxs+=view.getBorder.getMaxX
				else maxs+=view.getBorder.getMaxY
			}else {
				sortedSh.insert(i, view)
				mins.insert(i, coord)
				if(_distribution==DistributionType.horizEq)
					 maxs.insert(i, view.getBorder.getMaxX)
				else maxs.insert(i, view.getBorder.getMaxY)
			}
		}

		var gap = mins.last - maxs.head

		for(i <- 1 until sortedSh.size-1)
			gap -= maxs(i) - mins(i)

		gap/=(sortedSh.size-1)

		if(DistributionType.isVertical(_distribution))
			for(i <- 1 until sortedSh.size-1) {
				sortedSh(i).getShape.translate(0, (sortedSh(i-1).getBorder.getMaxY + gap) - mins(i))
				sortedSh(i).updateBorder
			}
		else
			for(i <- 1 until sortedSh.size-1) {
				sortedSh(i).getShape.translate((sortedSh(i-1).getBorder.getMaxX + gap) - mins(i), 0)
				sortedSh(i).updateBorder
			}

		sortedSh.clear
		mins.clear
		maxs.clear
	}


	/**
	 * Distributes using bottom/top/left/right reference.
	 */
	protected def distributeNotEq() {
		var sortedSh= new ListBuffer[IViewShape]()
		var centres	= new ListBuffer[Double]()
		var i : Int = 0
		val shapes = shape.get.getShapes

		_views.foreach{view =>
			val x = _distribution match {
				case DistributionType.horizLeft => view.getBorder.getMinX
				case DistributionType.horizMid => view.getBorder.getCenterX
				case DistributionType.horizRight => view.getBorder.getMaxX
				case DistributionType.vertBot => view.getBorder.getMaxY
				case DistributionType.vertMid => view.getBorder.getCenterY
				case DistributionType.vertTop => view.getBorder.getMinY
			}

			i = centres.indexWhere{value=> x<value}

			if(i == -1) {
				sortedSh+=view
				centres+=x
			}else {
				sortedSh.insert(i, view)
				centres.insert(i, x)
			}
		}

		val gap = (centres.last - centres.head)/(_views.size-1)

		if(DistributionType.isVertical(_distribution))
			for(i <- 1 until sortedSh.size-1)
				sortedSh(i).getShape.translate(0, (centres.head+i*gap)-centres(i))
		else
			for(i <- 1 until sortedSh.size-1)
				sortedSh(i).getShape.translate((centres.head+i*gap)-centres(i), 0)

		sortedSh.clear
		centres.clear
	}


	override def redo() {
		if(_distribution==DistributionType.horizEq || _distribution==DistributionType.vertEq)
			distributeEq
		else
			distributeNotEq
		shape.get.setModified(true)
	}


	/**
	 * Sets the alignment to perform.
	 */
	def setDistribution(distribution:DistributionType) {
		_distribution = distribution
	}

	/**
	 * Sets the reference border that must bounds the shapes to align.
	 */
	def setBorder(rec:Rectangle2D) {
		_border = rec
	}

	override def getUndoName() = "distribute"

	override def isRegisterable()  = true

	override def flush() {
		super.flush
		_views.clear
		_oldPositions.clear
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/java/net/sf/latexdraw/actions/shape/DistributeShapes.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2017 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.actions.shape;

import java.util.ArrayList;
import java.util.List;
import java.util.OptionalInt;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.util.LangTool;
import net.sf.latexdraw.view.jfx.Canvas;
import net.sf.latexdraw.view.jfx.ViewShape;

/**
 * This action distributes the provided shapes.
 * @author Arnaud Blouin
 */
public class DistributeShapes extends AlignDistribAction {
	/**
	 * This enumeration describes the different possible alignment types.
	 */
	public enum Distribution {
		VERT_BOT, VERT_TOP, VERT_MID, VERT_EQ, HORIZ_LEFT, HORIZ_RIGHT, HORIZ_MID, HORIZ_EQ;

		public static boolean isVertical(final Distribution distrib) {
			return distrib == VERT_BOT || distrib == VERT_TOP || distrib == VERT_MID || distrib == VERT_EQ;
		}
	}

	/** The views corresponding to the shapes to align. */
	private List<ViewShape<?>> views;
	/** The alignment to perform. */
	private Distribution distribution;
	private Canvas canvas;

	public DistributeShapes() {
		super();
	}

	@Override
	protected void doActionBody() {
		views = shape.get().getShapes().stream().map(sh -> canvas.getViewFromShape(sh).get()).collect(Collectors.toList());
		oldPositions = shape.get().getShapes().stream().map(sh -> sh.getTopLeftPoint()).collect(Collectors.toList());
		redo();
	}

	@Override
	public boolean canDo() {
		return shape.isPresent() && !shape.get().isEmpty() && canvas != null && distribution != null;
	}

	/**
	 * Distributes at equal distance between the shapes.
	 */
	private void distributeEq() {
		final List<ViewShape<?>> sortedSh = new ArrayList<>();
		final List<Double> mins = new ArrayList<>();
		final List<Double> maxs = new ArrayList<>();

		for(final ViewShape<?> view : views) {
			final double coord = distribution == Distribution.HORIZ_EQ ? view.getBoundsInLocal().getMinX() : view.getBoundsInLocal().getMinY();
			final OptionalInt res = IntStream.range(0, mins.size()).filter(index -> coord < mins.get(index)).findFirst();

			if(res.isPresent()) {
				final int i = res.getAsInt();
				sortedSh.add(i, view);
				mins.add(i, coord);
				maxs.add(i, distribution == Distribution.HORIZ_EQ ? view.getBoundsInLocal().getMaxX() : view.getBoundsInLocal().getMaxY());
			}else {
				sortedSh.add(view);
				mins.add(coord);
				maxs.add(distribution == Distribution.HORIZ_EQ ? view.getBoundsInLocal().getMaxX() : view.getBoundsInLocal().getMaxY());
			}
		}

		double gap = mins.get(mins.size() - 1) - maxs.get(0);

		for(int i = 1, size = sortedSh.size() - 1; i < size; i++) {
			gap -= maxs.get(i) - mins.get(i);
		}

		gap /= sortedSh.size() - 1;
		final double finalGap = gap;

		if(Distribution.isVertical(distribution)) {
			IntStream.range(1, sortedSh.size() - 1).forEach(i -> ((IShape) sortedSh.get(i).getUserData()).
				translate(0d, sortedSh.get(i - 1).getBoundsInLocal().getMaxY() + finalGap - mins.get(i)));
		}else {
			IntStream.range(1, sortedSh.size() - 1).forEach(i -> ((IShape) sortedSh.get(i).getUserData()).
				translate(sortedSh.get(i - 1).getBoundsInLocal().getMaxX() + finalGap - mins.get(i), 0d));
		}
	}

	/**
	 * Distributes using bottom/top/left/right reference.
	 */
	private void distributeNotEq() {
		final List<ViewShape<?>> sortedSh = new ArrayList<>();
		final List<Double> centres = new ArrayList<>();

		for(final ViewShape<?> view : views) {
			double x = 0;
			switch(distribution) {
				case HORIZ_LEFT:
					x = view.getBoundsInLocal().getMinX();
					break;
				case HORIZ_MID:
					x = (view.getBoundsInLocal().getMinX() + view.getBoundsInLocal().getMaxX()) / 2d;
					break;
				case HORIZ_RIGHT:
					x = view.getBoundsInLocal().getMaxX();
					break;
				case VERT_BOT:
					x = view.getBoundsInLocal().getMaxY();
					break;
				case VERT_MID:
					x = (view.getBoundsInLocal().getMinY() + view.getBoundsInLocal().getMaxY()) / 2d;
					break;
				case VERT_TOP:
					x = view.getBoundsInLocal().getMinY();
					break;
			}

			final double finalX = x;
			final OptionalInt res = IntStream.range(0, centres.size()).filter(index -> finalX < centres.get(index)).findFirst();

			if(res.isPresent()) {
				final int i = res.getAsInt();
				sortedSh.add(i, view);
				centres.add(i, x);
			}else {
				sortedSh.add(view);
				centres.add(x);
			}
		}

		final double gap = (centres.get(centres.size() - 1) - centres.get(0)) / (views.size() - 1);

		if(Distribution.isVertical(distribution)) {
			IntStream.range(1, sortedSh.size() - 1).forEach(i -> ((IShape) sortedSh.get(i).getUserData()).
				translate(0d, centres.get(0) + i * gap - centres.get(i)));
		}else {
			IntStream.range(1, sortedSh.size() - 1).forEach(i -> ((IShape) sortedSh.get(i).getUserData()).
				translate(centres.get(0) + i * gap - centres.get(i), 0d));
		}
	}

	@Override
	public void redo() {
		if(distribution == Distribution.HORIZ_EQ || distribution == Distribution.VERT_EQ) {
			distributeEq();
		}else {
			distributeNotEq();
		}
		shape.ifPresent(sh -> sh.setModified(true));
	}

	/**
	 * Sets the alignment to perform.
	 */
	public void setDistribution(final Distribution distrib) {
		distribution = distrib;
	}

	public void setCanvas(final Canvas theCanvas) {
		canvas = theCanvas;
	}

	@Override
	public String getUndoName() {
		return LangTool.INSTANCE.getBundle().getString("Actions.6");
	}

	@Override
	public RegistrationPolicy getRegistrationPolicy() {
		return RegistrationPolicy.LIMITED;
	}
}
